File: cslavb/Csla/Reflection/UndoableHandler.vb
Imports System
Imports System.Collections.Generic
Imports System.Reflection

Namespace Reflection
  NotInheritable Class UndoableHandler
    Private Sub New()
    End Sub
    Private Shared ReadOnly _undoableFieldCache As New Dictionary(Of Type, List(Of DynamicMemberHandle))()

    Public Shared Function GetCachedFieldHandlers(ByVal type As Type) As List(Of DynamicMemberHandle)
      Dim handlers As List(Of DynamicMemberHandle)
      If Not _undoableFieldCache.TryGetValue(type, handlers) Then
        Dim newHandlers = BuildHandlers(type)
        SyncLock _undoableFieldCache
          'ready to add, lock 
          If Not _undoableFieldCache.TryGetValue(type, handlers) Then
            _undoableFieldCache.Add(type, newHandlers)
            handlers = newHandlers
          End If
        End SyncLock
      End If
      Return handlers
    End Function

    Private Shared Function BuildHandlers(ByVal type As Type) As List(Of DynamicMemberHandle)
      Dim handlers = New List(Of DynamicMemberHandle)()
      ' get the list of fields in this type 
      Dim fields = type.GetFields(BindingFlags.NonPublic Or BindingFlags.Instance Or BindingFlags.[Public])

      For Each field As FieldInfo In fields
        ' make sure we process only our variables 
        If field.DeclaringType Is type Then
          ' see if this field is marked as not undoable 
          If Not NotUndoableField(field) Then
            ' the field is undoable, so it needs to be processed. 
            handlers.Add(New DynamicMemberHandle(field))
          End If
        End If
      Next
      Return handlers
    End Function

    Private Shared Function NotUndoableField(ByVal field As FieldInfo) As Boolean
      Return Attribute.IsDefined(field, GetType(NotUndoableAttribute))
    End Function

  End Class
End Namespace

File: Source/Csla.Shared/Core/UndoableHandler.cs
#if !NETFX_CORE && !IOS
//-----------------------------------------------------------------------
// <copyright file="UndoableHandler.cs" company="Marimer LLC">
//     Copyright (c) Marimer LLC. All rights reserved.
//     Website: https://cslanet.com
// </copyright>
// <summary>no summary</summary>
//-----------------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Reflection;

#if NET5_0_OR_GREATER
using System.Runtime.Loader;

using Csla.ALC;
#endif

using Csla.Reflection;

namespace Csla.Core
{
    static class UndoableHandler
    {
#if NET5_0_OR_GREATER
        private static readonly Dictionary<Type, Tuple<string, List<DynamicMemberHandle>>> _undoableFieldCache = 
          new Dictionary<Type, Tuple<string,  List<DynamicMemberHandle>>>();
#else
    private static readonly Dictionary<Type, List<DynamicMemberHandle>> _undoableFieldCache = new Dictionary<Type, List<DynamicMemberHandle>>();
#endif

        public static List<DynamicMemberHandle> GetCachedFieldHandlers(Type type)
        {
            List<DynamicMemberHandle> handlers;

#if NET5_0_OR_GREATER
            var found = _undoableFieldCache.TryGetValue(type, out var handlersInfo);

            handlers = handlersInfo?.Item2;

            if (!found)
            {

                lock (_undoableFieldCache)
                {
                    found = _undoableFieldCache.TryGetValue(type, out handlersInfo);

                    handlers = handlersInfo?.Item2;

                    if (!found)
                    {
                        var newHandlers = BuildHandlers(type);

                        var cacheInstance = ALCManager.CreateCacheInstance(type, newHandlers, OnAssemblyLoadContextUnload);

                        _undoableFieldCache.Add(type, cacheInstance);

                        handlers = newHandlers;
                    }
                }
            }
#else
            if (!_undoableFieldCache.TryGetValue(type, out handlers))
            {
              var newHandlers = BuildHandlers(type);
              lock (_undoableFieldCache)  //ready to add, lock
              {
                if (!_undoableFieldCache.TryGetValue(type, out handlers))
                {
                  _undoableFieldCache.Add(type, newHandlers);
                  handlers = newHandlers;
                }
              }
            }
#endif

            return handlers;
        }

        private static List<DynamicMemberHandle> BuildHandlers(Type type)
        {
            var handlers = new List<DynamicMemberHandle>();
            // get the list of fields in this type
            var fields = type.GetFields(
                BindingFlags.NonPublic |
                BindingFlags.Instance |
                BindingFlags.Public);

            foreach (FieldInfo field in fields)
            {
                // make sure we process only our variables
                if (field.DeclaringType == type)
                {
                    // see if this field is marked as not undoable 
                    if (!NotUndoableField(field))
                    {
                        // the field is undoable, so it needs to be processed.
                        handlers.Add(new DynamicMemberHandle(field));
                    }
                }
            }
            return handlers;
        }

        private static bool NotUndoableField(FieldInfo field)
        {
            // see if this field is marked as not undoable or IsInitOnly (ie: readonly property)
            return field.IsInitOnly || Attribute.IsDefined(field, typeof(NotUndoableAttribute));
        }

#if NET5_0_OR_GREATER
        private static void OnAssemblyLoadContextUnload(AssemblyLoadContext context)
        {
            lock (_undoableFieldCache)
                ALCManager.RemoveFromCache(_undoableFieldCache, context);
        }
#endif
    }
}
#endif
